src/util/fs.cc

```diff
@@ -121,4 +131,460 @@ bool os_portable_copy_file_to_dir(const std::string& src,
   return ret == 0;
 }
 
+/// Extract path from path
+/// C:\a\b\c.txt -> "C:\a\b\" 
+/// C:\a\b\c -> C:\a\b
+/// d/e/ghi  -> d/e/
+std::string os_portable_path_from_path(const std::string& path) {
+
+  std::string sep;
+#if defined(_WIN32) || defined(_WIN64)
+  // on windows
+  sep = "\\";
+#else
+  // on *nix
+  sep = "/";
+#endif
+  if (path.find(sep) == path.npos)
+    return "./";
+  auto pos = path.rfind(sep);
+  ILA_ASSERT(pos != path.npos);
+  return path.substr(0,pos+1); // include sep
+}
+
+/// C:\\a.txt -> C:\\a   or  /a/b/c.txt -> a/b/c
+std::string os_portable_remove_file_name_extension(const std::string& fname) {
+  std::string sep;
+#if defined(_WIN32) || defined(_WIN64)
+  // on windows
+  sep = "\\";
+#else
+  // on *nix
+  sep = "/";
+#endif
+  auto dot_pos = fname.rfind('.');
+  if(dot_pos == std::string::npos)
+    return fname; // no dot
+  
+  auto sep_pos = fname.rfind(sep);
+  
+  if (sep_pos == std::string::npos) // no sep and only dot
+    return fname.substr(0,dot_pos); // remove after dot
+
+  // no change ./../a -> ./../a
+  if(dot_pos < sep_pos)
+    return fname;
+
+  // /.asdfaf.d -> /.asdfaf  | /.a -> /.
+  return fname.substr(0,dot_pos);
+}
+
+
+#if defined(__unix__) || defined(unix) || defined(__APPLE__) || defined(__MACH__) || defined(__linux__) || defined(__FreeBSD__)
+
+volatile sig_atomic_t child_pid;
+volatile sig_atomic_t shared_time_out;
+
+void parent_alarm_handler(int signum) {
+  if (child_pid != 0) {
+    kill(-child_pid, SIGTERM);
+    shared_time_out = 1;
+  }
+}
+#endif
+
+#if defined(_WIN32) || defined(_WIN64)
+  #if defined(_MSC_VER) || defined(_MSC_EXTENSIONS)
+    #define DELTA_EPOCH_IN_MICROSECS  11644473600000000Ui64
+  #else
+    #define DELTA_EPOCH_IN_MICROSECS  11644473600000000ULL
+  #endif
+
+  struct timezone 
+  {
+    int  tz_minuteswest; /* minutes W of Greenwich */
+    int  tz_dsttime;     /* type of dst correction */
+  };
+
+  int gettimeofday(struct timeval *tv, struct timezone *tz)
+  {
+    FILETIME ft;
+    unsigned __int64 tmpres = 0;
+    static int tzflag;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The scope of the variable 'tzflag' can be reduced.](https://app.codacy.com/app/Bo-Yuan-Huang/ILAng/pullRequest?prid=4555809)